### PR TITLE
fix DeepEqual method for different types

### LINKED ISSUE #7386: Seed Reconciliation issues in weekly-2021-27
**User Story**
When  I try to upgrade the existing kkp installation to weekly-2021-27, my seed does not get reconciled and kubermatic-operator pod spews out continuous warnings (with log-debug=true). 

Note the first line of log below.. It talks about ENV being empty array vs not being present. This difference seems to be ignorable. But after this line, no further log comes for like a second and then it dumps the error that it timed out waiting for reconciliation.

```
{"level":"debug","time":"2021-07-14T11:46:09.528Z","caller":"reconciling/compare.go:60","msg":"Object differs from generated one","type":"*v1.Deployment","namespace":"kubermatic","name":"kubermatic-seed-controller-manager","diff":["Spec.Template.Spec.Containers.slice[0].Env: [] != <nil slice>"]}
{"level":"error","time":"2021-07-14T11:46:10.109Z","logger":"kubermatic-seed-operator","caller":"seed/reconciler.go:74","msg":"failed to reconcile","request":"kubermatic/vsphere-seed","error":"failed to reconcile Kubermatic Deployments: failed to ensure Deployment kubermatic/kubermatic-seed-controller-manager: failed waiting for the cache to contain our latest changes: timed out waiting for the condition"}
{"level":"error","time":"2021-07-14T11:46:10.110Z","logger":"controller-runtime.manager.controller.kubermatic-seed-operator","caller":"controller/controller.go:252","msg":"Reconciler error","name":"vsphere-seed","namespace":"kubermatic","error":"failed to reconcile Kubermatic Deployments: failed to ensure Deployment kubermatic/kubermatic-seed-controller-manager: failed waiting for the cache to contain our latest changes: timed out waiting for the condition"}
{"level":"debug","time":"2021-07-14T11:46:10.109Z","logger":"controller-runtime.manager.events","caller":"record/event.go:308","msg":"Warning","object":{"kind":"KubermaticConfiguration","namespace":"kubermatic","name":"kubermatic","uid":"bf809a0c-c492-4271-b30e-5ca236231e2e","apiVersion":"operator.kubermatic.io/v1alpha1","resourceVersion":"52819763"},"reason":"SeedReconcilingError","message":"vsphere-seed: failed to reconcile Kubermatic Deployments: failed to ensure Deployment kubermatic/kubermatic-seed-controller-manager: failed waiting for the cache to contain our latest changes: timed out waiting for the condition"}
```

If I rollback to v2.17.2 - these warnings disappear and seed gets reconciled properly.

As Lovro pointed out.. it could be due to.. [this change](https://github.com/kubermatic/kubermatic/pull/7220/files#diff-bdb313299c08d80f20d961410fd103663692455b59c08f7bb1c77abeb563c781R41)

The reconciling stops after it fails here, so the resources that are next in line never get reconciled.

**Acceptance criteria**
The compare should ignore this difference in ENV and should reconcile the seed properly.

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: fix DeepEqual method for different types

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7386

```release-note
NONE
```

### CODE REVIEW

**Comment by zreigz** [comment]
@irozzo-1A approve needed from you

File: pkg/resources/reconciling/compare.go [hunk 1]
```diff
@@ -38,8 +40,17 @@ func init() {
 
 // DeepEqual compares both objects for equality
 func DeepEqual(a, b metav1.Object) bool {
-	if reflect.DeepEqual(a, b) {
-		return true
+	// The ClusterTemplate consist with ClusterSpec which has nested Semver type.
+	// The semverlib.Version doesn't implement Equality function. Using equality.Semantic.DeepEqual for this object cause
+	// an informative panic.
+	if _, ok := a.(*kubermaticv1.ClusterTemplate); ok {
```
Comment (irozzo-1A) [inline pkg/resources/reconciling/compare.go]: @zreigz BTW a more elegant solution would be to create our `Equalities` instance providing a comparison function for `Semver` type. See the following for reference:

https://github.com/kubernetes/apimachinery/blob/b63a0c883fbfc313249150449400788e5589ef23/pkg/api/equality/semantic.go#L29-L49

Comment (zreigz) [reply pkg/resources/reconciling/compare.go]: I will try, thanks

Comment (lsviben) [reply pkg/resources/reconciling/compare.go]: I tested it and you can add this to the init function of `compare.go`
Or we can add this to the Add method of the ClusterTemplate controller, but then it only works for the controller, and in future maybe something else will need it. 

```
func init() {
	// Kubernetes Objects can be deeper than the default 10 levels.
	deep.MaxDepth = 20
	deep.LogErrors = true

	// Add comparison function for semver.Version
	equality.Semantic.AddFunc(func(a, b *semver.Version) bool {
		return a.Compare(b) == 0
	})
}

```

Comment (zreigz) [reply pkg/resources/reconciling/compare.go]: I had to add other objects to make it run our controller, seems to work fine
Thanks, @irozzo-1A

**Comment by zreigz** [comment]
@irozzo-1A @lsviben PTAL

**Comment by lsviben** [comment]
Looks like this solution is easily extendable if we need it. 
/lgtm

**Comment by xrstf** [comment]
I cancelled the test, just to test if restarting it will generate proper URLs. Sorry for the spam.

/retest

**Pull request merged**